File: src/系统/08．任务系统/04．任务UI拆分/02．任务UI辅助.lua
local ____lualib = require("lualib_bundle")
local __TS__StringSubstring = ____lualib.__TS__StringSubstring
local __TS__StringCharCodeAt = ____lualib.__TS__StringCharCodeAt
local __TS__StringCharAt = ____lualib.__TS__StringCharAt
local __TS__ParseInt = ____lualib.__TS__ParseInt
local __TS__ArrayFilter = ____lualib.__TS__ArrayFilter
local __TS__ArraySlice = ____lualib.__TS__ArraySlice
local __TS__ArraySome = ____lualib.__TS__ArraySome
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign
local __TS__ArrayMap = ____lualib.__TS__ArrayMap
local __TS__ArraySort = ____lualib.__TS__ArraySort
local ____exports = {}
local ____01_FF0E_4EFB_52A1UI_5E38_91CF = require("系统.08．任务系统.04．任务UI拆分.01．任务UI常量")
local TASK_UI_TOC_LOAD_KEY = ____01_FF0E_4EFB_52A1UI_5E38_91CF.TASK_UI_TOC_LOAD_KEY
local TASK_UI_TOC_PATHS = ____01_FF0E_4EFB_52A1UI_5E38_91CF.TASK_UI_TOC_PATHS
local ENABLE_FDF_A = ____01_FF0E_4EFB_52A1UI_5E38_91CF.ENABLE_FDF_A
local ENABLE_FDF_B = ____01_FF0E_4EFB_52A1UI_5E38_91CF.ENABLE_FDF_B
local ENABLE_FDF_SCROLLBAR = ____01_FF0E_4EFB_52A1UI_5E38_91CF.ENABLE_FDF_SCROLLBAR
local ENABLE_FDF_SCROLLBAR_BORDER = ____01_FF0E_4EFB_52A1UI_5E38_91CF.ENABLE_FDF_SCROLLBAR_BORDER
local ENABLE_FDF_SCROLLBAR_THUMB = ____01_FF0E_4EFB_52A1UI_5E38_91CF.ENABLE_FDF_SCROLLBAR_THUMB
local ____index = require("系统.09．表现系统.01．UI工具.index")
local loadTocOnce = ____index.loadTocOnce
local ____01_FF0E_4EFB_52A1_6570_636E = require("系统.08．任务系统.01．任务数据")
local questDB = ____01_FF0E_4EFB_52A1_6570_636E.questDB
local QuestType = ____01_FF0E_4EFB_52A1_6570_636E.QuestType
local QuestStatus = ____01_FF0E_4EFB_52A1_6570_636E.QuestStatus
local jass = require("jass.common")
local japi = require("jass.japi")
local __dzPcallShowJapi = nil
local __dzPcallShowFrame = 0
local __dzPcallShowVis = false
local function __dzPcallFrameShowBody(self)
    __dzPcallShowJapi:DzFrameShow(__dzPcallShowFrame, __dzPcallShowVis)
end
function ____exports.pcallDzFrameShow(self, japiAny, frame, visible)
    if type(japiAny.DzFrameShow) ~= "function" then
        return
    end
    __dzPcallShowJapi = japiAny
    __dzPcallShowFrame = frame
    __dzPcallShowVis = visible
    pcall(__dzPcallFrameShowBody)
    __dzPcallShowJapi = nil
end
local __dzPcallAlphaVal = 0
local function __dzPcallFrameSetAlphaBody(self)
    __dzPcallShowJapi:DzFrameSetAlpha(__dzPcallShowFrame, __dzPcallAlphaVal)
end
function ____exports.pcallDzFrameSetAlpha(self, japiAny, frame, alpha)
    if type(japiAny.DzFrameSetAlpha) ~= "function" then
        return
    end
    __dzPcallShowJapi = japiAny
    __dzPcallShowFrame = frame
    __dzPcallAlphaVal = alpha
    pcall(__dzPcallFrameSetAlphaBody)
    __dzPcallShowJapi = nil
end
function ____exports.dzGetLocalPlayer(self)
    return jass:GetLocalPlayer()
end
function ____exports.dzPlayer(self, index)
    return jass:Player(index)
end
function ____exports.questIdTailInRange01to20(self, id, prefix)
    if #id ~= #prefix + 3 then
        return false
    end
    if __TS__StringSubstring(id, 0, #prefix) ~= prefix then
        return false
    end
    local tail = __TS__StringSubstring(id, #prefix)
    if #tail ~= 3 then
        return false
    end
    return tail >= "001" and tail <= "020"
end
function ____exports.isQuestWithRowIconLayout(self, quest)
    if quest.icon and quest.icon ~= "" then
        return true
    end
    local id = quest.id
    if quest.type == QuestType.MAIN then
        return ____exports.questIdTailInRange01to20(nil, id, "main_")
    end
    if quest.type == QuestType.SIDE then
        return ____exports.questIdTailInRange01to20(nil, id, "side_")
    end
    if quest.type == QuestType.DAILY then
        return ____exports.questIdTailInRange01to20(nil, id, "daily_")
    end
    return true
end
function ____exports.isFdfFrameEnabled(self, frameName)
    local isA = frameName == "TaskEntryIcon" or frameName == "TaskMainPanel" or frameName == "TaskListContainer"
    local isB = frameName == "TaskTabMain" or frameName == "TaskTabSide" or frameName == "TaskTabDaily" or frameName == "TaskButtonBackdrop" or frameName == "TaskTabMainBg" or frameName == "TaskTabSideBg" or frameName == "TaskTabDailyBg"
    if frameName == "TaskScrollBar" then
        return ENABLE_FDF_SCROLLBAR
    end
    if frameName == "TaskScrollBarBorder" then
        return ENABLE_FDF_SCROLLBAR_BORDER
    end
    if frameName == "TaskScrollThumb" then
        return ENABLE_FDF_SCROLLBAR_THUMB
    end
    if isA then
        return ENABLE_FDF_A
    end
    if isB then
        return ENABLE_FDF_B
    end
    return false
end
local function tryCreateFromFdfOnlyNullFallback(self)
    return nil
end
local __dzCreateName = ""
local __dzCreateParent = 0
local __dzCreateContextId = 0
local __dzCreateResultFrame = 0
local function __dzCreateFramePcallBody(self)
    __dzCreateResultFrame = japi:DzCreateFrame(__dzCreateName, __dzCreateParent, __dzCreateContextId)
end
function ____exports.tryCreateFromFdfWithSource(self, name, parent, fallback, contextId)
    if contextId == nil then
        contextId = 0
    end
    if not ____exports.isFdfFrameEnabled(nil, name) then
        return {
            frame = fallback(nil),
            fromFdf = false
        }
    end
    loadTocOnce(nil, TASK_UI_TOC_LOAD_KEY, TASK_UI_TOC_PATHS, "TaskUI")
    __dzCreateName = name
    __dzCreateParent = parent
    __dzCreateContextId = contextId
    local ok = pcall(__dzCreateFramePcallBody)
    local f = __dzCreateResultFrame
    if ok and f ~= nil and f ~= 0 then
        return {frame = f, fromFdf = true}
    end
    return {
        frame = fallback(nil),
        fromFdf = false
    }
end
function ____exports.tryCreateFromFdfOnly(self, name, parent, contextId)
    if contextId == nil then
        contextId = 0
    end
    local res = ____exports.tryCreateFromFdfWithSource(
        nil,
        name,
        parent,
        tryCreateFromFdfOnlyNullFallback,
        contextId
    )
    if res.fromFdf and res.frame and res.frame ~= 0 then
        return res.frame
    end
    return nil
end
function ____exports.getStatusText(self, status)
    local m = {
        [QuestStatus.IN_PROGRESS] = "|cff00ff66进行中|r",
        [QuestStatus.COMPLETED] = "|cffc0c0c0已完成|r",
        [QuestStatus.FAILED] = "|cffff5555已失败|r",
        [QuestStatus.DISCOVERED] = "|cff66ccff已发现|r",
        [QuestStatus.UNDISCOVERED] = "|cff888888未发现|r"
    }
    return m[status] or status
end
local function questIdTailIsAllDigits(self, s)
    if #s == 0 then
        return false
    end
    do
        local i = 0
        while i < #s do
            local c = __TS__StringCharCodeAt(s, i)
            if c < 48 or c > 57 then
                return false
            end
            i = i + 1
        end
    end
    return true
end
--- TSTL 无 `lastIndexOf`，手写从右找 `_`
local function lastUnderscoreIndex(self, s)
    do
        local i = #s - 1
        while i >= 0 do
            if __TS__StringCharAt(s, i) == "_" then
                return i
            end
            i = i - 1
        end
    end
    return -1
end
--- `foo_2` 与 `foo_10` 字典序会乱；同一「末段 `_` 前」前缀且尾为纯数字时按数值比，否则字典序
local function compareQuestIdForListOrder(self, aId, bId)
    local ua = lastUnderscoreIndex(nil, aId)
    local ub = lastUnderscoreIndex(nil, bId)
    if ua > 0 and ub > 0 then
        local preA = __TS__StringSubstring(aId, 0, ua + 1)
        local preB = __TS__StringSubstring(bId, 0, ub + 1)
        if preA == preB then
            local tailA = __TS__StringSubstring(aId, ua + 1)
            local tailB = __TS__StringSubstring(bId, ub + 1)
            if questIdTailIsAllDigits(nil, tailA) and questIdTailIsAllDigits(nil, tailB) then
                local na = __TS__ParseInt(tailA, 10)
                local nb = __TS__ParseInt(tailB, 10)
                if na ~= nb then
                    return na - nb
                end
            end
        end
    end
    if aId < bId then
        return -1
    end
    if aId > bId then
        return 1
    end
    return 0
end
function ____exports.getQuestsForUI(self, playerId, ____type)
    local active = __TS__ArrayFilter(
        questDB:getPlayerActiveQuests(playerId),
        function(____, q) return q.type == ____type and not q.uiReserved end
    )
    local completedIds = questDB:getPlayerCompletedQuests(playerId)
    local result = __TS__ArraySlice(active)
    for ____, id in ipairs(completedIds) do
        do
            local template = questDB:getQuest(id)
            if not template or template.type ~= ____type or template.uiReserved then
                goto __continue51
            end
            if __TS__ArraySome(
                active,
                function(____, q) return q.id == id end
            ) then
                goto __continue51
            end
            result[#result + 1] = __TS__ObjectAssign(
                {},
                template,
                {
                    status = QuestStatus.COMPLETED,
                    objectives = __TS__ArrayMap(
                        template.objectives,
                        function(____, o) return __TS__ObjectAssign({}, o, {completed = true, current = o.required}) end
                    )
                }
            )
        end
        ::__continue51::
    end
    __TS__ArraySort(
        result,
        function(____, a, b) return compareQuestIdForListOrder(nil, a.id, b.id) end
    )
    return result
end
____exports.EMPTY_TEXTS = {[QuestType.MAIN] = "暂无主线任务", [QuestType.SIDE] = "暂无支线任务", [QuestType.DAILY] = "暂无小任务"}
return ____exports

File: TS/系统/08．任务系统/04．任务UI拆分/02．任务UI辅助.ts
import {
  TASK_UI_TOC_LOAD_KEY,
  TASK_UI_TOC_PATHS,
  ENABLE_FDF_A,
  ENABLE_FDF_B,
  ENABLE_FDF_SCROLLBAR,
  ENABLE_FDF_SCROLLBAR_BORDER,
  ENABLE_FDF_SCROLLBAR_THUMB,
} from "./01．任务UI常量";
import { loadTocOnce } from "../../09．表现系统/01．UI工具/index";
import { questManager } from "../02．任务管理器/index";
import { questDB, QuestType, QuestStatus, QuestData } from "../01．任务数据";

const jass = require("jass.common") as any;
const japi = require("jass.japi") as any;

// ── Dz 调用 pcall：仅用模块顶层具名体 + 全局 `pcall(具名)`（勿 `(pcall as any)(具名)`，TSTL 会编成 `pcall(nil, fn)`） ──
let __dzPcallFrame = 0;
let __dzPcallVis = false;

function __dzPcallFrameShowBody(): void {
  japi.DzFrameShow(__dzPcallFrame, __dzPcallVis);
}

export function pcallDzFrameShow(frame: number, visible: boolean): void {
  __dzPcallFrame = frame;
  __dzPcallVis = visible;
  pcall(__dzPcallFrameShowBody);
}

let __dzPcallAlphaVal = 0;

function __dzPcallFrameSetAlphaBody(): void {
  japi.DzFrameSetAlpha(__dzPcallFrame, __dzPcallAlphaVal);
}

export function pcallDzFrameSetAlpha(frame: number, alpha: number): void {
  __dzPcallFrame = frame;
  __dzPcallAlphaVal = alpha;
  pcall(__dzPcallFrameSetAlphaBody);
}

export function dzGetLocalPlayer(): any {
  return jass.GetLocalPlayer();
}

export function dzPlayer(index: number): any {
  return jass.Player(index);
}

export function questIdTailInRange01to20(id: string, prefix: string): boolean {
  if (id.length !== prefix.length + 3) return false;
  if (id.substring(0, prefix.length) !== prefix) return false;
  const tail = id.substring(prefix.length);
  if (tail.length !== 3) return false;
  return tail >= "001" && tail <= "020";
}

export function isQuestWithRowIconLayout(quest: QuestData): boolean {
  if (quest.icon && quest.icon !== "") return true;
  const id = quest.id;
  if (quest.type === QuestType.MAIN) return questIdTailInRange01to20(id, "main_");
  if (quest.type === QuestType.SIDE) return questIdTailInRange01to20(id, "side_");
  if (quest.type === QuestType.DAILY) return questIdTailInRange01to20(id, "daily_");
  // 任务 UI 默认给列表项保留图标位，避免测试任务/运行时任务因为 id 命名不同而整列丢图标。
  return true;
}

export function isFdfFrameEnabled(frameName: string): boolean {
  const isA = frameName === "TaskEntryIcon" || frameName === "TaskMainPanel" || frameName === "TaskListContainer";
  const isB =
    frameName === "TaskTabMain" ||
    frameName === "TaskTabSide" ||
    frameName === "TaskTabDaily" ||
    frameName === "TaskButtonBackdrop" ||
    frameName === "TaskTabMainBg" ||
    frameName === "TaskTabSideBg" ||
    frameName === "TaskTabDailyBg";
  if (frameName === "TaskScrollBar") return ENABLE_FDF_SCROLLBAR;
  if (frameName === "TaskScrollBarBorder") return ENABLE_FDF_SCROLLBAR_BORDER;
  if (frameName === "TaskScrollThumb") return ENABLE_FDF_SCROLLBAR_THUMB;
  if (isA) return ENABLE_FDF_A;
  if (isB) return ENABLE_FDF_B;
  return false;
}

function tryCreateFromFdfOnlyNullFallback(): number | null {
  return null;
}

let __dzCreateName = "";
let __dzCreateParent = 0;
let __dzCreateContextId = 0;
let __dzCreateResultFrame = 0;

function __dzCreateFramePcallBody(): void {
  __dzCreateResultFrame = (japi as any).DzCreateFrame(__dzCreateName, __dzCreateParent, __dzCreateContextId);
}

export function tryCreateFromFdfWithSource(
  name: string,
  parent: number,
  fallback: () => number | null,
  contextId: number = 0
): { frame: number | null; fromFdf: boolean } {
  if (!isFdfFrameEnabled(name)) return { frame: fallback(), fromFdf: false };
  loadTocOnce(TASK_UI_TOC_LOAD_KEY, TASK_UI_TOC_PATHS, "TaskUI");
  __dzCreateName = name;
  __dzCreateParent = parent;
  __dzCreateContextId = contextId;
  // void 具名体写槽位 + `local ok = pcall(具名)`：首返回值即 boolean，勿对 `pcall` 结果做下标/解构元组（见 `01．帧创建` 同构实现）。
  const ok = pcall(__dzCreateFramePcallBody);
  const f = __dzCreateResultFrame;
  if (ok && f != null && f !== 0) return { frame: f, fromFdf: true };
  return { frame: fallback(), fromFdf: false };
}

export function tryCreateFromFdfOnly(name: string, parent: number, contextId: number = 0): number | null {
  const res = tryCreateFromFdfWithSource(name, parent, tryCreateFromFdfOnlyNullFallback, contextId);
  if (res.fromFdf && res.frame && res.frame !== 0) return res.frame;
  return null;
}

export function getStatusText(status: QuestStatus): string {
  const m: Record<string, string> = {
    [QuestStatus.IN_PROGRESS]: "|cff00ff66进行中|r",
    [QuestStatus.COMPLETED]: "|cffc0c0c0已完成|r",
    [QuestStatus.FAILED]: "|cffff5555已失败|r",
    [QuestStatus.DISCOVERED]: "|cff66ccff已发现|r",
    [QuestStatus.UNDISCOVERED]: "|cff888888未发现|r",
  };
  return m[status] || status;
}

function questIdTailIsAllDigits(s: string): boolean {
  if (s.length === 0) return false;
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i);
    if (c < 48 || c > 57) return false;
  }
  return true;
}

/** TSTL 无 `lastIndexOf`，手写从右找 `_` */
function lastUnderscoreIndex(s: string): number {
  for (let i = s.length - 1; i >= 0; i--) {
    if (s.charAt(i) === "_") return i;
  }
  return -1;
}

/** `foo_2` 与 `foo_10` 字典序会乱；同一「末段 `_` 前」前缀且尾为纯数字时按数值比，否则字典序 */
function compareQuestIdForListOrder(aId: string, bId: string): number {
  const ua = lastUnderscoreIndex(aId);
  const ub = lastUnderscoreIndex(bId);
  if (ua > 0 && ub > 0) {
    const preA = aId.substring(0, ua + 1);
    const preB = bId.substring(0, ub + 1);
    if (preA === preB) {
      const tailA = aId.substring(ua + 1);
      const tailB = bId.substring(ub + 1);
      if (questIdTailIsAllDigits(tailA) && questIdTailIsAllDigits(tailB)) {
        const na = parseInt(tailA, 10);
        const nb = parseInt(tailB, 10);
        if (na !== nb) return na - nb;
      }
    }
  }
  if (aId < bId) return -1;
  if (aId > bId) return 1;
  return 0;
}

export function getQuestsForUI(playerId: number, type: QuestType): QuestData[] {
  const active = questDB
    .getPlayerActiveQuests(playerId)
    .filter(q => q.type === type && !q.uiReserved);
  const completedIds = questDB.getPlayerCompletedQuests(playerId);
  const result: QuestData[] = active.slice();

  for (const id of completedIds) {
    const template = questDB.getQuest(id);
    if (!template || template.type !== type || template.uiReserved) continue;
    if (active.some(q => q.id === id)) continue;
    result.push({
      ...template,
      status: QuestStatus.COMPLETED,
      objectives: template.objectives.map(o => ({ ...o, completed: true, current: o.required })),
    });
  }

  // 稳定顺序：避免底层表迭代顺序在各端不一致 → chunk 分页数/重绘次数分叉 → 联机不同步
  result.sort((a, b) => compareQuestIdForListOrder(a.id, b.id));
  return result;
}

export const EMPTY_TEXTS: Record<QuestType, string> = {
  [QuestType.MAIN]: "暂无主线任务",
  [QuestType.SIDE]: "暂无支线任务",
  [QuestType.DAILY]: "暂无小任务",
};
